pkg/resourcevalidator/common.go

```diff
@@ -0,0 +1,117 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcevalidator
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	klog "k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+	fcutils "github.com/liqotech/liqo/pkg/utils/foreignCluster"
+	"github.com/liqotech/liqo/pkg/virtualKubelet/forge"
+)
+
+var cachelog = klog.LoggerWithName(klog.Background(), "[ webhook-cache ]")
+var shadowpodlog = klog.LoggerWithName(klog.Background(), "[ shadowpod-resource ]")
+var webhooklog = klog.LoggerWithName(klog.Background(), "[ webhook ]")
+var resourceofferlog = klog.LoggerWithName(klog.Background(), "[ resourceoffer-resource ]")
+var webhookrefreshlog = klog.LoggerWithName(klog.Background(), "[ webhook-refresh ]")
+
+// DecodeShadowPod decodes a shadow pod from a given runtime object.
+func (spv *ShadowPodValidator) DecodeShadowPod(obj runtime.RawExtension) (shadowpod *vkv1alpha1.ShadowPod, err error) {
+	shadowpod = &vkv1alpha1.ShadowPod{}
+	err = spv.decoder.DecodeRaw(obj, shadowpod)
+	return
+}
+
+func quotaFormatter(quota v1.ResourceList, quotaName string) string {
+	result := fmt.Sprintf("%s [ ", quotaName)
+	r := regexp.MustCompile("^hugepages-")
+	for k, v := range quota {
+		if res := r.MatchString(k.String()); !res {
+			result += fmt.Sprintf("%s: %s, ", k, v.String())
+		}
+	}
+	result += "]"
+	return result
+}
+
+func getQuotaFromShadowPod(shadowpod *vkv1alpha1.ShadowPod, validate bool) (v1.ResourceList, error) {
+	resources := v1.ResourceList{}
+	for _, container := range shadowpod.Spec.Pod.Containers {
+		if container.Resources.Limits != nil {
+			for key, value := range container.Resources.Limits {
+				if prevResources, ok := resources[key]; ok {
+					prevResources.Add(value)
+					resources[key] = prevResources
+				} else {
+					resources[key] = value.DeepCopy()
+				}
+			}
+		} else if validate {
+			return nil, fmt.Errorf("container %s has no resource limits defined", container.Name)
+		}
+
+	}
+
+	for _, initContainer := range shadowpod.Spec.Pod.InitContainers {
+		if initContainer.Resources.Limits != nil || !validate {
+			for key, value := range initContainer.Resources.Limits {
+				if prevResources, ok := resources[key]; ok {
+					prevResources.Add(value)
+					resources[key] = prevResources
+				} else {
+					resources[key] = value.DeepCopy()
+				}
+			}
```

Comment:
so can i remove the entire "for" cycle for the initContainers?

Reply:
No, you should to adapt the logic as for the document above. IIRC, it should be something along the lines of max(sum(container_resources), max(init_container_resources)).

Reply:
1. In case of the init containers doesn't define any limit for CPU or Memory i will return an error as defined for the containers?
2. max(init_container_resources) means thats i have to select the maximum between initContainers and not the sum of them? if yes, which criteria i have to use to select one between InitContainer 1 (CPU: 1, Memory: 2) and InitContainer 2 (CPU: 2, Memory: 1), who "wins"? 
maybe it is related to the maximum of each type of resource type, right (in this case the result is CPU: 2, Memory: 2)?

Reply:
I would try to match as much as possible the behavior of standard resource quotas. You should double check, but I think the correct approach is:
1. returning an error
2. considering the max for each resource type

Reply:
perfect! Thank you